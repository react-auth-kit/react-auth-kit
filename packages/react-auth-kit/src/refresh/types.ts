/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Payload for Refresh token False
 */
type RefreshTokenActionResponseFailed = {
  /**
   * If the refresh operation is successful or not
   *
   * If the `isSuccess` is `true`, then the `token` and other items will be
   * replaced with the new network response
   *
   * If the isSuccess is `false`, then everything will be wiped and
   * the user will be signed out
   */
  isSuccess: false
}

/**
 * Payload for Refresh token Success True
 */
export type RefreshTokenActionResponsePassed<T> = {

  /**
   * New Auth token from the network response
   */
  newAuthToken: string,

  /**
   * New Auth Token type from the network response
   */
  newAuthTokenType?: string,

  /**
   * New Refresh token from the network response. Can be null
   */
  newRefreshToken?: string,

  /**
   * New User state from the network. Can be null
   */
  newAuthUserState?: T | null,

  /**
   * If the refresh operation is successful or not
   *
   * If the isSuccess is `true`, then the `token` and other items will be
   * replaced with the new network response
   *
   * If the isSuccess is `false`, then everything will be wiped and the user will
   * be signed out
   */
  isSuccess: boolean
}

/**
 * Response type by the refresh token
 */
export type RefreshTokenActionResponse<T> = RefreshTokenActionResponsePassed<T> | RefreshTokenActionResponseFailed;

/**
 * Callback function for the refresh token operation
 *
 * This function is called when the refresh token operation is triggered.
 * It should return a promise which resolves to a `RefreshTokenActionResponse`
 * object.
 *
 * @typeParam T - Type of User State Object
 */
export type refreshTokenCallback<T> = (param: {

  /**
   * Existing Auth token for the refresh operation
   */
  authToken?: string,

  /**
   * Existing Refresh token for the refresh operation
   */
  refreshToken?: string,

  /**
   * Existing User State for the User state
   */
  authUserState: T | null,
}) => Promise<RefreshTokenActionResponse<T>>
