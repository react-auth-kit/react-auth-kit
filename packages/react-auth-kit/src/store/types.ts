/*
 * Copyright  $year Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type {createRefreshAttribute} from "../refresh";
import type {ITokenStore} from "./ITokenStore";

/**
 * Return type of authStore Function
 */
export interface Store<T> {
  /**
   * Instance of the token object
   */
  tokenStore: ITokenStore<T>;
  /**
   * Instance of the Refresh interface, if there is any.
   */
  refresh?: createRefreshAttribute<T>;
}

/**
 * Set State Data
 */
export interface UpdatedAuthKitState<T> {

  /**
   * Authentication Object
   */
  auth?: {

    /**
     * JWT access token
     */
    token: string,

    /**
     * Type of the access token
     *
     * @example
     * Bearer
     */
    type: string
  } | null,

  /**
   * Refresh JWT token
   */
  refresh?: string | null,

  /**
   * User state object
   */
  userState?: T
}
