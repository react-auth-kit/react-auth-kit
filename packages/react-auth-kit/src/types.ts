/**
 * @author Arkadip Bhattacharya <in2arkadipb13@gmail.com>
 * @fileoverview General types used in the library
 * @copyright Arkadip Bhattacharya 2020
 *
 * Copyright 2020 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {AuthActions} from './utils/actions';

/**
 * Universal types
 */
export type AuthStateUserObject = {
  [x: string]: any;
}

export interface AuthKitStateInterface {
  auth: {
    token: string,
    type: string,
    expiresAt: Date
  } | null,
  refresh: {
    token: string,
    expiresAt: Date
  } | null,
  userState: AuthStateUserObject | null,
  isSignIn: boolean
  isUsingRefreshToken: boolean,
  // typeOfStorage: "cookie" | "localstorage"
}

/**
 * SignIn function param
 * Used by: useSignIn and withSignIn
 */
export interface signInFunctionParams {
  token: string
  tokenType: string | 'Bearer'
  expiresIn: number
  authState?: AuthStateUserObject
  refreshToken?: string
  refreshTokenExpireIn?: number
}

/**
 * Context values type
 */
export interface AuthContextInterface {
  authState: AuthKitStateInterface
  dispatch: React.Dispatch<AuthActions>
}

/**
 * Auth Provider Props
 */
export interface AuthProviderProps {
  authType: 'cookie' | 'localstorage'
  authName: string,
  refresh?: createRefreshParamInterface
  cookieDomain?: string
  cookieSecure?: boolean
  children: React.ReactNode
}

/**
 * Refresh Token Callback Response
 */
export type RefreshTokenCallbackResponse = {
  isSuccess: boolean,
  newAuthToken: string,
  newAuthTokenExpireIn?: number | null,
  newRefreshToken?: string | null,
  newRefreshTokenExpiresIn?: number | null,
  newAuthUserState?: AuthStateUserObject | null,
};

/**
 * Refresh Token types
 */
// Callback function

export type refreshTokenCallback = (param: {
  authToken?: string,
  authTokenExpireAt?: Date,
  refreshToken?: string,
  refreshTokenExpiresAt?: Date,
  authUserState: AuthStateUserObject | null,
}) => Promise<RefreshTokenCallbackResponse>

// createRefresh function parameter
export interface createRefreshParamInterface {
  interval: number,
  refreshApiCallback: refreshTokenCallback
}

export type refreshFunctionType = (param: createRefreshParamInterface)
  => createRefreshParamInterface
