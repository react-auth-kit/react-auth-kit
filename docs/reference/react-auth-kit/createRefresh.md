---
title: createRefresh - React Auth Kit
description: Create the Refresh Token engine for React Auth Kit
---


# createRefresh

<div data-ea-publisher="authkitarkadipme" data-ea-type="text" id="ref_createRefresh"></div>

## Import

```js
import createRefresh from 'react-auth-kit/createRefresh';
```


## Function Signature

**createRefresh**<`T`\>([`createRefreshParamInterface`](./types.md#createrefreshparaminterface)): [`createRefreshParamInterface`](./types.md#createrefreshparaminterface)<`T`\>

## Type Parameters

| Name | Description               |
|:-----|:--------------------------|
| `T`  | Type of User State Object |

## Parameters

| Name    | Type                                                                          | Required           | Description                                |
|:--------|:------------------------------------------------------------------------------|:-------------------|:-------------------------------------------|
| `param` | [`createRefreshParamInterface`](./types.md#createrefreshparaminterface)<`T`\> | :heavy_check_mark: | Parameters required for the refresh engine |

## Returns

[`createRefreshParamInterface`]()<`T`\>

Same params with an added layer of safety net.

## Example

```js
const refresh = createRefresh({
    interval: 10,
    refreshApiCallback: async (param) => {
        try {
            const response = await axios.post("/refresh", param, {
                headers: {'Authorization': `Bearer ${param.authToken}`}
            })
            console.log("Refreshing")
            return {
                isSuccess: true,
                newAuthToken: response.data.token,
                newAuthTokenExpireIn: 10,
                newRefreshTokenExpiresIn: 60
            }
        }
        catch(error){
            console.error(error)
            return {
                isSuccess: false
            }
        }
    }
})

```

## Remarks

This function doesn't really "do anything" at runtime, it just helps to organize the code base
Use this function to create the refresh token system

---

#### Defined in

[packages/react-auth-kit/src/createRefresh.ts:119](https://github.com/react-auth-kit/react-auth-kit/blob/37dc30d4/packages/react-auth-kit/src/createRefresh.ts#L119)

<p align="center">&mdash; ðŸ”‘  &mdash;</p>
<p align="center"><i>React Auth Kit is <a href="https://github.com/react-auth-kit/react-auth-kit/blob/master/LICENSE">MIT License</a> code</i></p>
