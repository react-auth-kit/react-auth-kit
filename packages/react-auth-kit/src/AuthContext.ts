'use client';

import type {Context} from 'react';
import {createContext, useContext} from 'react';

import type Router from './route';
import type {ITokenStore} from "./store";
import {AuthKitProviderMissingError} from "./error";

interface ReactAuthKitContextConfig {
  fallbackPath?: string
}

interface ReactAuthKitContext<T> {
  store: ITokenStore<T>
  router?: Router
  config: ReactAuthKitContextConfig
}

/**
 * @internal
 * @returns React Context with Token Object inside
 *
 * React Context to globally hold the TokenObject instance in the application.
 *
 */
function getContext<T>(): Context<ReactAuthKitContext<T>> {
  const context = createContext<ReactAuthKitContext<T>>(null as any);
  if (process.env.NODE_ENV !== 'production') {
    context.displayName = 'ReactAuthKit';
  }
  return context;
}

/**
 * @internal
 * @returns React Context with Token Object inside
 *
 * Use this context to globally hold the TokenObject instance in the application.
 *
 */
const AuthKitContext = getContext();

/**
 *
 * @internal
 * @returns TokenObject from the context
 *
 * React Context consumer to globally hold the
 * TokenObject instance in the application.
 */
export function useReactAuthKitStore(): ITokenStore<unknown> {
  const context = useContext(AuthKitContext);
  if (context === null) {
    throw new
    AuthKitProviderMissingError(
        'Auth Provider is missing. ' +
        'Make sure, you are using this component inside the auth provider.',
    );
  }
  return context.store;
}

/**
 *
 * @internal
 * @returns Router Object from the context
 *
 * React Context consumer to globally hold the
 * Router instance in the application.
 */
export function useReactAuthKitRouter(): Router|undefined {
  const context = useContext(AuthKitContext);
  if (context === null) {
    throw new
    AuthKitProviderMissingError(
        'Auth Provider is missing. ' +
        'Make sure, you are using this component inside the auth provider.',
    );
  }
  return context.router;
}

/**
 * @internal
 * @returns React Auth Kit configurations
 *
 * React Context consumer to globally hold the
 * React Auth Kit configurations in the application.
 */
export function useReactAuthKitConfig(): ReactAuthKitContextConfig {
  const context = useContext(AuthKitContext);
  if (context === null) {
    throw new
    AuthKitProviderMissingError(
        'Auth Provider is missing. ' +
        'Make sure, you are using this component inside the auth provider.',
    );
  }
  return context.config;
}

export default AuthKitContext;
