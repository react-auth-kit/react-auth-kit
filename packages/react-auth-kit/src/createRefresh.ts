/*
 * Copyright 2020 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {createRefreshParamInterface} from './types';

/**
 * This function doesn't really "do anything" at runtime,
 * it's just help to organize the code base
 * Use this function to create the refresh token system
 */
function createRefresh(param: createRefreshParamInterface)
  :createRefreshParamInterface {
  return param;
}

export default createRefresh;
