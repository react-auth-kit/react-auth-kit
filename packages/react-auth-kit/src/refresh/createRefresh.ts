/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use client';

import type React from 'react';

import {AuthKitConfigError} from "../error";
import {refreshTokenCallback} from "./types";


/**
 * Parameter for the Refresh operation
 */
export interface createRefreshAttribute<T> {

  /**
   * Interval on which the callback function is called
   */
  interval: number;

  /**
   * A Callback function which have the network request
   *
   * @example
   * ```js
   * refreshApiCallback: async (param) => {
   *  try {
   *    const response = await axios.post("/refresh", param, {
   *      headers: {'Authorization': `Bearer ${param.authToken}`}
   *    })
   *    console.log("Refreshing")
   *    return {
   *      isSuccess: true,
   *      newAuthToken: response.data.token,
   *      newAuthTokenExpireIn: 10,
   *      newRefreshTokenExpiresIn: 60
   *    }
   *  }
   *  catch(error){
   *    console.error(error)
   *    return {
   *      isSuccess: false
   *    }
   *  }
   * }
   * ```
   */
  refreshApiCallback: refreshTokenCallback<T>;

  /**
   * Initial component to be rendered when the refresh is not yet started
   * or the refresh is in progress.
   *
   * @remarks
   * This component is rendered only once when the refresh is started
   * and not at every refresh interval.
   * This is useful to show a loading spinner or a message
   *
   * @example
   * ```jsx
   * <Spinner />
   * ```
   */
  initialRefreshComponent?: React.ReactNode;
}

type createRefreshResponse<T> = createRefreshAttribute<T>

/**
 * @param param - Parameters required for the refresh engine
 * @returns Same params with added layer of a safety net.
 *
 * @remarks
 * This function doesn't really "do anything" at runtime,
 * it's just helps to organize the code base
 * Use this function to create the refresh token system
 *
 */
function createRefresh<T>(param: createRefreshAttribute<T>)
  :createRefreshResponse<T> {
  if (param.interval < 0) {
    throw new AuthKitConfigError(
        'Refresh interval is a time in seconds and can\'t be a negative(-ve)'+
        ' number. Make sure you are using positive number.',
    );
  }
  return param;
}

export default createRefresh;
