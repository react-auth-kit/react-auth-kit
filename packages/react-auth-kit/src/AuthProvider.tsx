import * as React from 'react';
import AuthContext from './AuthContext';
import TokenObject from './TokenObject';
import {AuthProviderProps} from './types';
import {authReducer, doRefresh} from './utils/reducers';
import {useInterval} from './utils/hooks';
import { AuthKitError } from './errors';


/**
 * AuthProvider - The Authentication Context Provider
 *
 * @param children
 * @param authStorageName
 * @param cookieDomain
 * @param cookieSecure
 *
 * @return Functional Component
 */
const AuthProvider: React.FunctionComponent<AuthProviderProps> =
  ({
    children,
    authType,
    authName,
    cookieDomain,
    cookieSecure,
    refresh,
  }) => {
    if (authType === 'cookie') {
      if (!cookieDomain) {
        throw new
        AuthKitError('authType \'cookie\' ' +
          'requires \'cookieDomain\' and \'cookieSecure\' ' +
          'props in AuthProvider');
      }
    }

    const refreshTokenName = refresh ? `${authName}_refresh` : null;

    const tokenObject = new TokenObject(authName, authType,
        refreshTokenName, cookieDomain, cookieSecure);

    const [authState, dispatch] =
      React.useReducer(authReducer, tokenObject.initialToken());

    if (refresh) {
      useInterval(
          () => {
            refresh
              .refreshApiCallback({
                authToken: authState.auth?.token,
                authTokenExpireAt: authState.auth?.expiresAt,
                authUserState: authState.userState,
                refreshToken: authState.refresh?.token,
                refreshTokenExpiresAt: authState.refresh?.expiresAt,
              })
              .then((result) => {
                // IF the API call is successful then refresh the AUTH state
                if (result.isSuccess) {
                  // store the new value using the state update
                  dispatch(doRefresh(result));
                }
                else {
                  // do something in future
                }
              })
              .catch(()=>{
                // do something in future
              });
          },
        authState.isSignIn ? refresh.interval : null,
      );
    }

    React.useEffect(() => {
      tokenObject.syncTokens(authState);
    }, [authState]);

    return (
      <AuthContext.Provider value={{authState, dispatch}}>
        {children}
      </AuthContext.Provider>
    );
  };

export default AuthProvider;
