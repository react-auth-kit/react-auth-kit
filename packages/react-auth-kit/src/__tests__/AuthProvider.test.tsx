/*
 * Copyright 2020 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {render, screen} from '@testing-library/react';
import {AuthContextConsumer} from '../AuthContext';
import AuthProvider from '../AuthProvider';
import {BrowserRouter, Route, Routes} from 'react-router-dom';
import React from 'react';
import createRefresh from '../createRefresh';
import {AuthStateUserObject} from '../types';
import Cookies from 'js-cookie';

describe('AuthProvider renders successfully', ()=>{
  it('With localStorage', ()=>{
    render(
        <AuthProvider authType={'localstorage'} authName={'_auth'}>
          <BrowserRouter>
            <Routes>
              <Route path={'/'} element={
                <div>
                  Empty
                </div>
              }/>
              <Route/>
            </Routes>
          </BrowserRouter>
        </AuthProvider>,
    );
  });
  describe('With cookie', ()=>{
    it('renders with all props given', ()=>{
      render(
          <AuthProvider
            authType={'cookie'}
            authName={'_auth'}
            cookieDomain={window.location.hostname}
            cookieSecure
          >
            <BrowserRouter>
              <Routes>
                <Route path={'/'} element={
                  <div>
                    Empty
                  </div>
                }/>
                <Route/>
              </Routes>
            </BrowserRouter>
          </AuthProvider>,
      );
    });

    it('Throws an error, then props are missing', ()=>{
      jest.spyOn(console, 'error').mockImplementation(jest.fn());

      expect(() => render(
          <AuthProvider
            authType={'cookie'}
            authName={'_auth'}
          >
            <BrowserRouter>
              <Routes>
                <Route path={'/'} element={
                  <div>
                    Empty
                  </div>
                }/>
                <Route/>
              </Routes>
            </BrowserRouter>
          </AuthProvider>,
      )).toThrow();
    });
  });
});
describe('Authprovider with refresh Token', ()=> {
  afterEach(() => {
    jest.useRealTimers();
  });
  const refreshApi = createRefresh({
    interval: 1/60,
    refreshApiCallback: (param): any => {
      console.log(param);
      const result = new Promise((resolve) => {
        return resolve({
          isSuccess: true,
          newAuthToken: 'fsdgedgd',
          newAuthTokenExpireIn: 10,
          newRefreshTokenExpiresIn: 60,
        });
      });
      return result;
    },
  });

  const refreshFalseApi = createRefresh({
    interval: 1/60,
    refreshApiCallback: (param): any => {
      console.log(param);
      const result = new Promise((resolve) => {
        return resolve({
          isSuccess: false,
        });
      });
      return result;
    },
  });

  const refreshApiWithError = createRefresh({
    interval: 1/60,
    refreshApiCallback: (param): any => {
      console.log(param);
      const result = new Promise((_resolve, reject) => {
        reject(new Error('Some error occured'));
      });
      return result;
    },
  });

  it('renders successfully', ()=>{
    const {container} = render(
        <AuthProvider
          authType={'localstorage'}
          authName={'_hi'}
          refresh={refreshApi}
        >
          <BrowserRouter>
            <Routes>
              <Route path={'/'} element={
                <div>
                  Empty
                </div>
              }/>
              <Route/>
            </Routes>
          </BrowserRouter>
        </AuthProvider>,
    );
    expect(container.nodeName).toMatch('DIV');
  });


  it('Initially signed in', ()=>{
    jest.useFakeTimers();
    const spyRefreshApi = jest.spyOn(refreshApi, 'refreshApiCallback');

    const authToken = '__authToken';
    const authType = '__authType';
    const authTime = new Date(2021, 10, 5);
    const refreshToken = '__refreshToken';
    const refreshTime = new Date(2021, 10, 6);
    const userState = {key: 'val'} as AuthStateUserObject;

    Cookies.get = jest
        .fn()
        .mockImplementationOnce(() => authToken)
        .mockImplementationOnce(() => authType)
        .mockImplementationOnce(() => authTime)
        .mockImplementationOnce(() => JSON.stringify(userState))
        .mockImplementationOnce(() => refreshToken)
        .mockImplementationOnce(() => refreshTime);

    const TestConsumer = () => <AuthContextConsumer>{
      (value) => (
        <span>
          isSignIn: {value?.authState.isSignIn ? 'true' : 'false'}
        </span>
      )
    }</AuthContextConsumer>;

    render(
        <AuthProvider
          authType={'cookie'}
          authName={'_hi'}
          cookieDomain={window.location.hostname}
          cookieSecure={window.location.protocol === 'https:'}
          refresh={refreshApi}
        >
          <BrowserRouter>
            <TestConsumer />
            <Routes>
              <Route path={'/'} element={
                <div>
                  Empty
                </div>
              }/>
              <Route/>
            </Routes>
          </BrowserRouter>
        </AuthProvider>,
    );

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).not.toBeCalled();
    jest.advanceTimersByTime(1000);

    // TODO: assertion statements needed
    expect(screen.getByText(/^isSignIn:/).textContent).toBe('isSignIn: true');

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).toHaveBeenCalled();

    // Check again if refreshApiCallback is called
    jest.advanceTimersByTime(1000);
    expect(spyRefreshApi).toHaveBeenCalledTimes(2);

    jest.useRealTimers();
  });

  it('Refresh API is giving False reponse', ()=>{
    jest.useFakeTimers();
    const spyRefreshApi = jest.spyOn(refreshFalseApi, 'refreshApiCallback');

    const authToken = '__authToken';
    const authType = '__authType';
    const authTime = new Date(2021, 10, 5);
    const refreshToken = '__refreshToken';
    const refreshTime = new Date(2021, 10, 6);
    const userState = {key: 'val'} as AuthStateUserObject;

    Cookies.get = jest
        .fn()
        .mockImplementationOnce(() => authToken)
        .mockImplementationOnce(() => authType)
        .mockImplementationOnce(() => authTime)
        .mockImplementationOnce(() => JSON.stringify(userState))
        .mockImplementationOnce(() => refreshToken)
        .mockImplementationOnce(() => refreshTime);

    const TestConsumer = () => <AuthContextConsumer>{
      (value) => (
        <span>
          isSignIn: {value?.authState.isSignIn ? 'true' : 'false'}
        </span>
      )
    }</AuthContextConsumer>;

    render(
        <AuthProvider
          authType={'cookie'}
          authName={'_hi'}
          cookieDomain={window.location.hostname}
          cookieSecure={window.location.protocol === 'https:'}
          refresh={refreshFalseApi}
        >
          <BrowserRouter>
            <TestConsumer />
            <Routes>
              <Route path={'/'} element={
                <div>
                  Empty
                </div>
              }/>
              <Route/>
            </Routes>
          </BrowserRouter>
        </AuthProvider>,
    );

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).not.toBeCalled();
    jest.advanceTimersByTime(1000);

    // TODO: assertion statements needed
    expect(screen.getByText(/^isSignIn:/).textContent).toBe('isSignIn: true');

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).toHaveBeenCalled();

    // Check again if refreshApiCallback is called
    jest.advanceTimersByTime(1000);
    expect(spyRefreshApi).toHaveBeenCalledTimes(2);

    jest.useRealTimers();
  });

  it('Refresh API is giving Error reponse', ()=>{
    jest.useFakeTimers();
    const spyRefreshApi = jest.spyOn(refreshApiWithError, 'refreshApiCallback');

    const authToken = '__authToken';
    const authType = '__authType';
    const authTime = new Date(2021, 10, 5);
    const refreshToken = '__refreshToken';
    const refreshTime = new Date(2021, 10, 6);
    const userState = {key: 'val'} as AuthStateUserObject;

    Cookies.get = jest
        .fn()
        .mockImplementationOnce(() => authToken)
        .mockImplementationOnce(() => authType)
        .mockImplementationOnce(() => authTime)
        .mockImplementationOnce(() => JSON.stringify(userState))
        .mockImplementationOnce(() => refreshToken)
        .mockImplementationOnce(() => refreshTime);

    const TestConsumer = () => <AuthContextConsumer>{
      (value) => (
        <span>
          isSignIn: {value?.authState.isSignIn ? 'true' : 'false'}
        </span>
      )
    }</AuthContextConsumer>;

    render(
        <AuthProvider
          authType={'cookie'}
          authName={'_hi'}
          cookieDomain={window.location.hostname}
          cookieSecure={window.location.protocol === 'https:'}
          refresh={refreshApiWithError}
        >
          <BrowserRouter>
            <TestConsumer />
            <Routes>
              <Route path={'/'} element={
                <div>
                  Empty
                </div>
              }/>
              <Route/>
            </Routes>
          </BrowserRouter>
        </AuthProvider>,
    );

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).not.toBeCalled();
    jest.advanceTimersByTime(1000);

    // TODO: assertion statements needed
    expect(screen.getByText(/^isSignIn:/).textContent).toBe('isSignIn: true');

    // Check if refreshApiCallback is called
    expect(spyRefreshApi).toHaveBeenCalled();

    // Check again if refreshApiCallback is called
    jest.advanceTimersByTime(1000);
    expect(spyRefreshApi).toHaveBeenCalledTimes(2);

    jest.useRealTimers();
  });
});

export {};
