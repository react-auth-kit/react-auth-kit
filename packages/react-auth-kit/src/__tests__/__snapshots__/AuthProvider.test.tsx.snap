// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider Renders Refresh component 1`] = `
<div>
  <div>
    <span
      id="refresh"
    >
      Refresh Token
    </span>
    <div
      id="test"
    >
      Hello
    </div>
  </div>
</div>
`;
